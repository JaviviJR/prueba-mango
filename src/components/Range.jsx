import PropTypes from 'prop-types'
import { useCallback, useEffect, useLayoutEffect, useRef, useState } from "react";
import './Range.styles.scss';
import RangeSlider from './RangeSlider';

function Range({
	min,
	max,
	// initialValue,
	onChange,
	width,
	// step = 5,
}) {
    const sliderRef = useRef(null);
    const [sliderWith, setSliderWidth] = useState(null);
    
    const minBulletRef = useRef(null);
    const [currentMinValue, setCurrentMinValue] = useState(min);
    
    const maxBulletRef = useRef(null);
    const [currentMaxValue, setCurrentMaxValue] = useState(max);
    
    const [isDragging, setIsDragging] = useState(false);
    const elementDragging = useRef(null);


    /* DEBUG */
    const [posX, setPosX] = useState(null);
    const [totalWidth, setTotalWidth] = useState(null);
    const [selectedValue, setSelectedValue] = useState(null);
    const [position, setPosition] = useState(null);

    const [minBulletMinState, setMinBulletMinState] = useState(null);
    const [minBulletMaxState, setMinBulletMaxState] = useState(null);
    const [maxBulletMinState, setMaxBulletMinState] = useState(null);
    const [maxBulletMaxState, setMaxBulletMaxState] = useState(null);
    /* END DEBUG */

    const moveSliderPosition = useCallback((e, bullet, setMethod) => {
        const sliderBoundingClientRect = sliderRef.current?.getBoundingClientRect();
        
        if (sliderBoundingClientRect) {
            const posX = (e.clientX) - sliderBoundingClientRect.left;
            setPosX(posX);

            const totalWidth = sliderBoundingClientRect.width;
            setSliderWidth(totalWidth);
            setTotalWidth(totalWidth);
            
            const {bulletMin, bulletMax} = getMinMaxValue(bullet);
            
            let selectedValue = Math.round((posX / totalWidth) * (max - min) + min);
            selectedValue = Math.max(bulletMin, selectedValue);
            selectedValue = Math.min(bulletMax, selectedValue);
            setSelectedValue(selectedValue);
            setMethod(selectedValue);
        }
    },
    [max, min, currentMinValue, currentMaxValue]);

    useEffect(() => {
        if (!minBulletRef.current) return;

        const {bulletMin, bulletMax} = getMinMaxValue(minBulletRef);
        setMinBulletMinState(bulletMin);
        setMinBulletMaxState(bulletMax);

        calculateBulletPosition(minBulletRef, currentMinValue);
    }, [currentMinValue]);

    useEffect(() => {
        if (!maxBulletRef.current) return;

        const { bulletMin, bulletMax } = getMinMaxValue(maxBulletRef);
        setMaxBulletMinState(bulletMin);
        setMaxBulletMaxState(bulletMax);

        calculateBulletPosition(maxBulletRef, currentMaxValue);
    }, [currentMaxValue]);

    const calculateBulletPosition = useCallback((bulletRef, value) => {
        const sliderBoundingClientRect = sliderRef.current?.getBoundingClientRect();
        const totalWidth = sliderBoundingClientRect.width;        
        const position = ((value - min) / (max - min)) * totalWidth;
        setPosition(position);
        bulletRef.current.style.left = `${position + (-1 * 20 / 2)}px`;
    }, [min, max]);

    const redrawBullets = useCallback(() => {
        calculateBulletPosition(minBulletRef, currentMinValue);
        calculateBulletPosition(maxBulletRef, currentMaxValue);
    }, [calculateBulletPosition, currentMinValue, currentMaxValue]);

    useLayoutEffect(() => {
        window.addEventListener("resize", redrawBullets);
        return () => {
			window.removeEventListener("resize", redrawBullets);
		};
    }, []);

    useEffect(() => {
        redrawBullets();
    }, [sliderWith,redrawBullets]);

    const getMinMaxValue = (bullet) => {
        switch (bullet) {
            case minBulletRef:
                return {
                    bulletMin: min,
                    bulletMax: currentMaxValue
                };
            case maxBulletRef:
                return {
                    bulletMin: currentMinValue,
                    bulletMax: max
                };
            default:
                console.log('default');
                break;
        }
    }

    const getBulletStateByElement = (element) => {
        switch (element) {
            case minBulletRef.current:
                return {
                    bulletRef: minBulletRef,
                    value: currentMinValue,
                    setMethod: setCurrentMinValue
                };
            case maxBulletRef.current:
                return {
                    bulletRef: maxBulletRef,
                    value: currentMaxValue,
                    setMethod: setCurrentMaxValue
                };
            default:
                return null;
        }
    };

    const onMouseUp = useCallback(() => {
		onChange({
            min: currentMinValue,
            max: currentMaxValue
        });
		setIsDragging(false);
        elementDragging.current = null;
	}, [currentMinValue, currentMaxValue, onChange]);

	const onMouseMove = useCallback(
		(e) => {
            const { bulletRef, setMethod } = getBulletStateByElement(elementDragging.current);
            if (isDragging) {
                moveSliderPosition(e, bulletRef, setMethod);
			}
		},
		[isDragging, moveSliderPosition]
	);

    const onMouseDown = (event) => {
        if (isDragging) return;

        const { bulletRef, setMethod } = getBulletStateByElement(event.target);
        
        if (!bulletRef.current) return;

		setIsDragging(true);
        elementDragging.current = bulletRef.current;
		moveSliderPosition(event, bulletRef, setMethod);
	};

    useEffect(() => {
		if (isDragging) {
			window.addEventListener("mousemove", onMouseMove);
			window.addEventListener("mouseup", onMouseUp);
		}

		return () => {
			window.removeEventListener("mousemove", onMouseMove);
			window.removeEventListener("mouseup", onMouseUp);
		};
	}, [isDragging, onMouseMove, onMouseUp]);

    return (
        <div>
            <RangeSlider
                min={min}
                max={max}
                width={width}
                currentMinValue={currentMinValue}
                setCurrentMinValue={setCurrentMinValue}
                currentMaxValue={currentMaxValue}
                isDragging={isDragging}
                onMouseDown={onMouseDown}
                // onTouchStart={onTouchStart}
                // onTouchStart={() => {}}
                // onKeyDown={onKeyDown}
                // onKeyDown={() => {}}
                sliderRef={sliderRef}
                minBulletRef={minBulletRef}
                maxBulletRef={maxBulletRef}
            />
            CurrentMin: { currentMinValue }<br />
            CurrentMax: { currentMaxValue }<br />
            posX: { posX }<br />
            sliderWith: { sliderWith }<br />
            totalWidth: { totalWidth }<br />
            selectedValue: { selectedValue }<br />
            position: { position }<br />
            minBulletMin: { minBulletMinState }<br />
            minBulletMax: { minBulletMaxState }<br />
            maxBulletMin: { maxBulletMinState }<br />
            maxBulletMax: { maxBulletMaxState }<br />
        </div>
    );
}

export default Range;

Range.propTypes = {
    min: PropTypes.number.isRequired,
    max: PropTypes.number.isRequired,
    // initialValue: PropTypes.number.isRequired,
    onChange: PropTypes.func.isRequired,
    width: PropTypes.number,
    // step: PropTypes.number,
};