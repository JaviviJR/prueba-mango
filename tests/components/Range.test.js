import { fireEvent, render, screen } from '@testing-library/react';
import Range from '../../src/components/Range';

describe('testinv <Range> compontent', () => {
    
    const onChangeMock = jest.fn();

    beforeEach(() => {
        jest.clearAllMocks();
    });
    
    test('should math with snapshot', () => {
        const { container } = render(
            <Range 
                mode={'normal'}
                min={0}
                max={100}
                onChange={onChangeMock}
            />
        );
        expect(container).toMatchSnapshot();
    })

    test('should have all elements', () => {
        render(
            <Range 
                mode={'normal'}
                min={0}
                max={100}
                onChange={onChangeMock}
            />
        );
        expect(screen.getAllByRole('button').length).toBe(2);
        expect(screen.getAllByRole('spinbutton').length).toBe(2);
        expect(screen.getAllByRole('slider').length).toBe(1);
    })

    test('debugeando', () => {
        // const user = userEvent.setup();

        const { container } = render(
            <Range 
                mode={'normal'}
                min={0}
                max={100}
                onChange={onChangeMock}
            />
        );
        
        const slider = screen.getByTestId('slider');
        const minBullet = screen.getByTestId('min-value-bullet');
        const maxBullet = screen.getByTestId('max-value-bullet');
        // user.click(minBullet);

        // jest.spyOn(slider, "getBoundingClientRect").mockImplementation(
        //     () => ({
        //             x: 0,
        //             y: 0,
        //             bottom: 0,
        //             height: 0,
        //             left: 0,
        //             right: 0,
        //             top: 0,
        //             width: 500
        // }));

        slider.getBoundingClientRect = jest.fn(
            () => ({
                    x: 0,
                    y: 0,
                    bottom: 0,
                    height: 10,
                    left: 0,
                    right: 0,
                    top: 0,
                    width: 500
        }));
        
        fireEvent.mouseDown(minBullet);
        fireEvent.mouseMove(container, {
            clientX: 1,
            clientY: 0,
        })

        fireEvent.mouseUp(minBullet)

        console.log('move');

        // user.pointer([
        //     // touch the screen at element1
        //     {keys: '[MouseLeft>]', target: minBullet},
        //     // move the touch pointer to element2
        //     {pointerName: 'MouseLeft', target: ?},
        //     // release the touch pointer at the last position (element2)
        //     {keys: '[/MouseLeft]'},
        // ]);

        console.log('slider', slider.getBoundingClientRect());
        console.log('min', minBullet.getBoundingClientRect());
        console.log('max', maxBullet.getBoundingClientRect());
        screen.debug();
    })
})