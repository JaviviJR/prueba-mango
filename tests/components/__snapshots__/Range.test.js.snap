// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testinv <Range> compontent should math with snapshot 1`] = `
<div>
  <div>
    <div
      class="range-slider-container"
      data-testid="slider"
      style="width: 100%;"
    >
      <div
        class="range-slider-container__input-div"
      >
        <input
          class="range-slider-container__input"
          data-testid="min-value-input"
          type="number"
          value="0.00"
        />
        €
      </div>
      <div
        class="range-slider"
        role="slider"
      >
        <div
          class="range-slider__bar"
        >
          <button
            class="range-slider__handle"
            data-testid="min-value-bullet"
            style="cursor: grab; left: -10px;"
          />
          <button
            class="range-slider__handle"
            data-testid="max-value-bullet"
            style="cursor: grab; left: -10px;"
          />
        </div>
      </div>
      <div
        class="range-slider-container__input-div"
      >
        <input
          class="range-slider-container__input"
          data-testid="max-value-input"
          type="number"
          value="100.00"
        />
        €
      </div>
    </div>
  </div>
</div>
`;
